### PR TITLE
Arrow functions, rest params, and template literals in routing and extension support

### PR DESCRIPTION
**Pull request by chadhietala**

### CODE REVIEW

**Comment by stefanpenner** [comment]
needs rebase

**Comment by chadhietala** [comment]
WAT. Looks like there is a bug in this JSCS thing.

```
One space required before closing curly brace at ember-routing/system/dsl.js :
    48 |
    49 |      createRoute(dsl, 'loading');
    50 |      createRoute(dsl, 'error', { path: `/_unused_dummy_error_path_route_${name}/:error` });
----------------------------------------------------------------------------------------^
    51 |
```

**Comment by rwjblue** [comment]
Not sure if it is a bug (likely `/_unused_dummy_error_path_route_${ name }/:error` would pass the test), but the assertion is being thrown from this rule:  https://github.com/jscs-dev/node-jscs/blob/master/lib/rules/require-spaces-inside-object-brackets.js.  You could ping the JSCS folks to see if they want to address?

**Comment by chadhietala** [comment]
@rwjblue Just tried that. Looks like JSCS is still not happy with your suggestion.

**Comment by rwjblue** [comment]
Ugh. So it wants a space after the curly (in the string)? If so, that is clearly not what we want and we will need to tweak/update the rule to fix...

```
`string${asdf} /asdf`
```

File: packages/ember-extension-support/lib/data_adapter.js [hunk 1]
```diff
@@ -132,22 +132,21 @@ export default EmberObject.extend({
...
       return wrapped;
     });
 
     typesAdded(typesToSend);
 
-    var release = function() {
-      releaseMethods.forEach(function(fn) { fn(); });
-      self.releaseMethods.removeObject(release);
+    var release = () => {
+      releaseMethods.forEach((fn) => { fn(); });
```
Comment (stefanpenner) [inline packages/ember-extension-support/lib/data_adapter.js]: `forEach((fn) => { fn(); })` => `forEach(fn) => fn())`

File: packages/ember-routing/lib/system/route.js [hunk 2]
```diff
@@ -143,14 +142,14 @@ var Route = EmberObject.extend(ActionHandler, Evented, {
       qps: qps,
       map: map,
       states: {
-        active(controller, prop) {
```
Comment (stefanpenner) [inline packages/ember-routing/lib/system/route.js]: we lost the `es6.properties.shorthand`

File: packages/ember-routing/lib/system/route.js [hunk 3]
```diff
@@ -1058,12 +1057,12 @@ var Route = EmberObject.extend(ActionHandler, Evented, {
     @param {String} name the name of the action to trigger
     @param {...*} args
   */
-  send() {
+  send(...args) {
     if (this.router || !Ember.testing) {
-      this.router.send(...arguments);
+      this.router.send.apply(this.router, args);
```
Comment (stefanpenner) [inline packages/ember-routing/lib/system/route.js]: `this.router.send(...args);`

**Comment by chadhietala** [comment]
Ok all attempts to fix this by later versions of JSCS have not worked and will continue not to work until Esprima is fixed and JSCS 1.12 is released. https://github.com/jscs-dev/node-jscs/issues/748.  @rwjblue or @stefanpenner I can wait or cat them together and come back this later.

**Comment by rwjblue** [comment]
@chadhietala you can disable individual rules for JSCS via code comments.  Something like the following:

```
// jscs:disable requireSpacesInsideObjectBrackets
`some${thing}/here`
//jscs:enable requireSpacesInsideObjectBrackets
```

**Comment by rwjblue** [comment]
I'm fine with either the comments, or just punting on the template strings for now and coming back to them when JSCS supports them.

**Comment by chadhietala** [comment]
@rwjblue went with option 3 :smile:. The error only gets thrown when you use a TemplateLiteral in an ObejectExpression. There are only a couple places in these files where we are building up a string in an ObjectExpression. See packages/ember-routing/lib/system/dsl.js in this commit.

All the jobs are passing except production and from the looks of it I'm not really sure why.

**Comment by rwjblue** [comment]
Failing test:

> Ember.View#element: should not allow the elementId to be changed after inserted


You can run locally by executing the following:

```
# normal way (includes all tests)
ember serve --environment=production

# exclude .min JSCS and JSHint tests -- MUCH Faster
DISABLE_ES3=true DISABLE_JSCS=true DISABLE_JSHINT=true DISABLE_MIN=true DISABLE_DEREQUIRE=true ember serve --environment=production
```

Then navigate to:

`http://localhost:4200/tests/index.html?skipPackage=container,ember-testing,ember-debug&dist=min&prod=true`

**Comment by stefanpenner** [comment]
looks like jscs will be released and fixed tomorrow:)

**Comment by hzoo** [comment]
1.12.0 is released now!

**Comment by chadhietala** [comment]
@stefanpenner, @rwjblue finally green.

**Pull request merged**